src/code/z_vimode.c

```diff
@@ -1,23 +1,378 @@
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/func_80140CE0.s")
+    yScaleHi0 = not_78 ? (cond_40 ? 0x1000000 : 0x2000000) : 0;
+    yScaleHi1 = not_78 ? (cond_40 ? 0x3000000 : 0x2000000) : 0;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/func_80140D04.s")
+    viMode->customViMode.type = mode;
+    viMode->customViMode.comRegs.ctrl = OS_VI_UNK2000 | OS_VI_UNK1000 | OS_VI_GAMMA | OS_VI_GAMMA_DITHER |
+                                        (!cond_44 ? OS_VI_UNK40 : 0) | (not_74 ? OS_VI_DIVOT : 0) |
+                                        (not_7C ? OS_VI_UNK2 | OS_VI_UNK1 : OS_VI_UNK2);
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/func_80140D10.s")
+    if (cond_3C) {
+        viMode->customViMode.comRegs.ctrl |= 0x100;
+    } else if (cond_38 | cond_34) {
+        viMode->customViMode.comRegs.ctrl |= 0x300;
+    } else if (unk_74) {
+        viMode->customViMode.comRegs.ctrl |= 0x200;
+    } else {
+        viMode->customViMode.comRegs.ctrl |= 0;
+    }
+
+    viMode->customViMode.comRegs.width = width * (cond_48 ? 2 : 1);
+
+    if (type < 3) {
+        static const ViModeStruct D_801DF0C0[] = {
+            {
+                0x0404233A,
+                0x00000270,
+                0x00150C69,
+                0x0C6F0C6E,
+                0x00800300,
+                0x005F0239,
+                0x0009026B,
+            },
+            {
+                0x03E52239,
+                0x0000020C,
+                0x00000C15,
+                0x0C150C15,
+                0x006C02EC,
+                0x002501FF,
+                0x000E0204,
+            },
+            {
+                0x04651E39,
+                0x0000020C,
+                0x00000C10,
+                0x0C1C0C1C,
+                0x006C02EC,
+                0x002501FF,
+                0x000E0204,
+            },
+        };
+        const ViModeStruct* ptr = &D_801DF0C0[type];
+
+        viMode->customViMode.comRegs.burst = ptr->unk_00;
+        viMode->customViMode.comRegs.vSync = ptr->unk_04;
+        viMode->customViMode.comRegs.hSync = ptr->unk_08;
+        viMode->customViMode.comRegs.leap = ptr->unk_0C;
+        viMode->customViMode.comRegs.hStart = ptr->unk_10;
+        viMode->customViMode.fldRegs[0].vStart = ptr->unk_14;
+        viMode->customViMode.fldRegs[1].vStart = ptr->unk_14;
+        viMode->customViMode.fldRegs[0].vBurst = ptr->unk_18;
+        viMode->customViMode.fldRegs[1].vBurst = ptr->unk_18;
+    } else {
+        __assert("../z_vimode.c", 216);
+    }
+
+    viMode->customViMode.comRegs.hStart += (unk_left << 16) + (s16)unk_right;
+    viMode->customViMode.fldRegs[0].vStart += (unk_top << 16) + (s16)unk_bottom;
+    viMode->customViMode.fldRegs[1].vStart += (unk_top << 16) + (s16)unk_bottom;
+
+    if (cond_44) {
+        viMode->customViMode.comRegs.vSync++;
+        if (type == 2) {
+            viMode->customViMode.comRegs.hSync += 0x40001;
+            viMode->customViMode.comRegs.leap += 0xFFFCFFFE;
```

Comment:
Is leap a signed number? I wonder if 0xFFFCFFFE is possibly negative? This seems like a strange number to add